pkg/gadgets/internal/test/helpers.go

```diff
@@ -0,0 +1,161 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/google/go-cmp/cmp"
+	"golang.org/x/sys/unix"
+)
+
+// CreateMntNsFilterMap creates and fills an eBPF map that can be used
+// to filter by mount namespace in the different tracers.
+func CreateMntNsFilterMap(t *testing.T, mountNsIDs ...uint64) *ebpf.Map {
+	t.Helper()
+
+	const one = uint32(1)
+
+	mntnsSpec := &ebpf.MapSpec{
+		Name:       "filter_map_" + t.Name(),
+		Type:       ebpf.Hash,
+		KeySize:    8,
+		ValueSize:  4,
+		MaxEntries: 1024,
+	}
+	m, err := ebpf.NewMap(mntnsSpec)
+	if err != nil {
+		t.Fatalf("Failed to create eBPF map: %s", err)
+	}
+
+	for _, mountnsid := range mountNsIDs {
+		if err := m.Put(mountnsid, one); err != nil {
+			m.Close()
+			t.Fatalf("Failed to update eBPF map: %s", err)
+		}
+	}
+
+	return m
+}
+
+// RequireRoot skips the test if the not running as root
+func RequireRoot(t *testing.T) {
+	t.Helper()
+	if unix.Getuid() != 0 {
+		t.Skip("Test requires root")
+	}
+}
+
+type Tracer interface {
+	Stop()
+}
+
+type ValidateEventType[Event any, Ret any] func(t *testing.T, info *Info, ret *Ret, events []Event)
+
+type Test[Event any, Ret any] struct {
+	GetTracer     func(*Info, func(event Event)) Tracer
+	RunnerConfig  *Config
+	GenerateEvent func() (Ret, error)
+	ValidateEvent ValidateEventType[Event, Ret]
+}
+
+// Run runs a series of test cases using a dedicated runner.
+func Run[Event any, Ret any](t *testing.T, tests map[string]Test[Event, Ret]) {
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			events := []Event{}
+			eventCallback := func(event Event) {
+				events = append(events, event)
+			}
+
+			runner, err := newRunner(test.RunnerConfig)
+			if err != nil {
+				t.Fatalf("%q: failed to create runner", err)
+			}
+			t.Cleanup(runner.close)
+
+			info := runner.Info
+
+			tracer := test.GetTracer(info, eventCallback)
+			t.Cleanup(tracer.Stop)
+
+			var ret Ret
+
+			f := func() error {
+				var err error
+				ret, err = test.GenerateEvent()
+				return err
+			}
+
+			if err := runner.run(f); err != nil {
+				t.Fatalf("Error generating event: %s", err)
+			}
+
+			// Give some time for the tracer to capture the events
+			time.Sleep(100 * time.Millisecond)
+
+			test.ValidateEvent(t, info, &ret, events)
+		})
+	}
+}
+
+// ExpectNoEvent doesn't expect any event to be captured by the tracer.
+func ExpectNoEvent[Event any, Ret any](t *testing.T, _ *Info, _ *Ret, events []Event) {
+	if len(events) != 0 {
+		t.Fatalf("no events are expected")
+	}
+}
+
+// ExpectAtLeastOneEvent expects that at least one of the captures events matches.
+func ExpectAtLeastOneEvent[Event any, Ret any](getEvent func(info *Info, ret *Ret) *Event) ValidateEventType[Event, Ret] {
```

Comment:
I still don't get why generic must be used and it seem to me they reduce readability a lot. I'd appreciate it you would find a moment to explain it to me. It's also the first time me reviewing code with generics :)

Reply:
The runner creates a tracer and saves all the events captured by it in an `event` slice o L84, then it calls `ValidateEvent` passing the captured events. Since the captured events are of a specific type (according to the tracer), the validate event must use generics to adapt to the different types.

Do you think there is another way to get this done without generics?

Reply:
If you only do `len()` and `DeepEqual()`, maybe just `[]interface{}` would work? :smile:

Reply:
Good question. I wasn't able to find a nice way to plug it without generics. IMO generics is ok to be used here as it avoids dealing with casting and interface{} all around.

Reply:
Fair enough. Marking as resolved then :)